cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,290 @@
+ * ENV{DEVTYPE}=="partition", ENV{ID_PART_ENTRY_NAME}="ubuntu-boot", SYMLINK+="disk/snapd/ubuntu-boot"
+ * ENV{DEVTYPE}=="partition", ENV{ID_PART_ENTRY_NAME}="ubuntu-data", ENV{ID_FS_TYPE}=="crypto_LUKS", SYMLINK+="disk/snapd/ubuntu-data-luks"
+ * ENV{DEVTYPE}=="partition", ENV{ID_PART_ENTRY_NAME}="ubuntu-data", ENV{ID_FS_TYPE}!="crypto_LUKS", SYMLINK+="disk/snapd/ubuntu-data"
+ * ENV{DEVTYPE}=="partition", ENV{ID_PART_ENTRY_NAME}="ubuntu-save", ENV{ID_FS_TYPE}=="crypto_LUKS", SYMLINK+="disk/snapd/ubuntu-save-luks"
+ * ENV{DEVTYPE}=="partition", ENV{ID_PART_ENTRY_NAME}="ubuntu-save", ENV{ID_FS_TYPE}!="crypto_LUKS", SYMLINK+="disk/snapd/ubuntu-save"
+ *
+ * LABEL="ubuntu_core_partitions_end"
+ *
+ * ENV{DM_UUID}=="CRYPT-*", ENV{DM_NAME}=="ubuntu-data-*", SYMLINK+="disk/snapd/ubuntu-data"
+ * ENV{DM_UUID}=="CRYPT-*", ENV{DM_NAME}=="ubuntu-save-*", SYMLINK+="disk/snapd/ubuntu-save"
+ * ```
+ *
+ * See
+ * core-initrd/latest/factory/usr/lib/udev/rules.d/90-ubuntu-core-partitions.rules
+ * for implementation.
+ *
+ * Note that symlink /dev/disk/snapd/disk can be expected by
+ * snap-bootstrap. In that case, snap-initramfs-mounts.service should
+ * have:
+ *
+ * ```
+ * BindsTo=dev-disk-snapd--disk.device
+ * After=dev-disk-snapd-disk.device
+ * ```
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/cmd/snap-bootstrap/blkid"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil/kcmdline"
+)
+
+func init() {
+	const (
+		short = "Verify that a disk is the booting disk"
+		long  = "This tool is expected to be called from udev"
+	)
+
+	addCommandBuilder(func(parser *flags.Parser) {
+		if _, err := parser.AddCommand("scan-disk", short, long, &cmdScanDisk{}); err != nil {
+			panic(err)
+		}
+	})
+}
+
+type cmdScanDisk struct{}
+
+func (c *cmdScanDisk) Execute([]string) error {
+	return ScanDisk(os.Stdout)
+}
+
+type Partition struct {
+	Name string
+	UUID string
+}
+
+func isGpt(probe blkid.AbstractBlkidProbe) bool {
+	pttype, err := probe.LookupValue("PTTYPE")
+	if err != nil {
+		return false
+	}
+	return pttype == "gpt"
+}
+
+func probePartitions(node string) ([]Partition, error) {
+	probe, err := blkid.NewProbeFromFilename(node)
+	if err != nil {
+		return nil, err
+	}
+	defer probe.Close()
+
+	probe.EnablePartitions(true)
+	probe.SetPartitionsFlags(blkid.BLKID_PARTS_ENTRY_DETAILS)
+	probe.EnableSuperblocks(true)
+
+	if err := probe.DoSafeprobe(); err != nil {
+		return nil, err
+	}
+
+	if !isGpt(probe) {
+		return nil, nil
+	}
+
+	partitions, err := probe.GetPartitions()
+	if partitions == nil {
+		return nil, err
+	}
+
+	ret := make([]Partition, 0)
+	for _, partition := range partitions.GetPartitions() {
+		label := partition.GetName()
+		uuid := partition.GetUUID()
+		fmt.Fprintf(os.Stderr, "Found partition %s %s\n", label, uuid)
```

Comment:
we need to be able to redirect this Fprintf so that the tests can capture them, see how cmd/snap deals with that for example

Reply:
Ah that might have been a debug line. I can remove that.

Reply:
In general a generator output goes to /dev/null. So we have to use /dev/kmsg if we really want to print something. So this was obviously debug.